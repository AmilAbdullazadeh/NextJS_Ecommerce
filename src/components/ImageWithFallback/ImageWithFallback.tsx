import React, { useState } from 'react';
import Image from 'next/image';

import { productImageUrl } from '@core';

export interface ImageFallbackProps {
    src: string;
    fallbackSrc: string;
    alt: string;
    width: number;
    height: number;
}

const ImageFallback: React.FC<ImageFallbackProps> = ({
    src,
    fallbackSrc,
    alt,
    width,
    height,
}: ImageFallbackProps) => {
    const [showFallback, setShowFallback] = useState(false);
    const [originalSrc, setOriginalSrc] = useState(src);

    const handleImgSrcError = () => setShowFallback(true);

    if (originalSrc !== src) {
        setShowFallback(false);
        setOriginalSrc(src);
    }

    return (
        <Image
            src={
                showFallback
                    ? productImageUrl(fallbackSrc)
                    : productImageUrl(src)
            }
            alt={alt}
            layout="responsive"
            width={width}
            height={height}
            onError={handleImgSrcError}
        />
    );
};

export default ImageFallback;
