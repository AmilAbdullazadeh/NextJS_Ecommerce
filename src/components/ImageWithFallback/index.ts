export { default as ImageWithFallback } from './ImageWithFallback';
