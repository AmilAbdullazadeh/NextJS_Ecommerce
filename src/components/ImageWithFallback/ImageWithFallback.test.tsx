import React from 'react';

import { render } from '@testUtils';

import ImageWithFallback, { ImageFallbackProps } from './ImageWithFallback';

describe('ImageWithFallback', () => {
    const defaultProps: ImageFallbackProps = {
        src: 'iphone11.png',
        fallbackSrc: 'phoneFallBack.png',
        alt: 'phone name',
        width: 200,
        height: 200,
    };

    it('renders correctly', () => {
        const {
            container: { firstChild },
        } = render(<ImageWithFallback {...defaultProps} />);

        expect(firstChild).toMatchSnapshot();
    });

    it('renders correctly for error img src', () => {
        const {
            container: { firstChild },
        } = render(<ImageWithFallback {...defaultProps} src="" />);

        expect(firstChild).toMatchSnapshot();
    });

    it('rerenders correctly when it receives a new image', () => {
        const {
            container: { firstChild },
            rerender,
            getByAltText,
            queryByAltText,
        } = render(<ImageWithFallback {...defaultProps} />);

        expect(firstChild).toMatchSnapshot();
        expect(getByAltText(defaultProps.alt)).toBeInTheDocument();

        rerender(
            <ImageWithFallback
                {...defaultProps}
                src="iphone13.jpg"
                alt="new phone name"
            />
        );

        expect(firstChild).toMatchSnapshot();
        expect(queryByAltText(defaultProps.alt)).not.toBeInTheDocument();
        expect(getByAltText('new phone name')).toBeInTheDocument();
    });
});
