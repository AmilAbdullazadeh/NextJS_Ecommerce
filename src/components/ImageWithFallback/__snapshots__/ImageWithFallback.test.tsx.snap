// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageWithFallback renders correctly 1`] = `
<div
  style="display: block; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
>
  <div
    style="display: block; box-sizing: border-box; padding-top: 100%;"
  />
  <noscript />
  <img
    alt="phone name"
    data-nimg="true"
    decoding="async"
    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
  />
</div>
`;

exports[`ImageWithFallback renders correctly for error img src 1`] = `
<div
  style="display: block; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
>
  <div
    style="display: block; box-sizing: border-box; padding-top: 100%;"
  />
  <noscript />
  <img
    alt="phone name"
    data-nimg="true"
    decoding="async"
    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
  />
</div>
`;

exports[`ImageWithFallback rerenders correctly when it receives a new image 1`] = `
<div
  style="display: block; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
>
  <div
    style="display: block; box-sizing: border-box; padding-top: 100%;"
  />
  <noscript />
  <img
    alt="phone name"
    data-nimg="true"
    decoding="async"
    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
  />
</div>
`;

exports[`ImageWithFallback rerenders correctly when it receives a new image 2`] = `
<div
  style="display: block; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
>
  <div
    style="display: block; box-sizing: border-box; padding-top: 100%;"
  />
  <noscript />
  <img
    alt="new phone name"
    data-nimg="true"
    decoding="async"
    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
  />
</div>
`;
